docs/spec/spec.md

```diff
@@ -0,0 +1,950 @@
+#
+# + anyValue - Any value to be unpacked
+# + targetTypeOfAny - Type descriptor of the return value
+# + return - Return a value of the given type
+public isolated function unpack(Any anyValue, ValueTypeDesc targetTypeOfAny = <>) returns targetTypeOfAny|Error;
+```
+
+# 4. gRPC Communication
+
+gRPC has 4 types of RPCs (Remote Procedure Calls), and Ballerina supports all of them.
+1. Unary
+2. Server streaming
+3. Client streaming
+4. Bidirectional streaming
+
+Note that, to explain the behaviour of these 4 RPC types, this document uses the standard Route Guide example.
+- [Details of the route guide example](https://grpc.io/docs/languages/go/basics/)
+- [Protocol buffer definition of the route guide example](https://github.com/ballerina-platform/module-ballerina-grpc/blob/674bda12a90f99c2735badc5567cd7dd7e14ba09/examples/routeguide/proto-file/route_guide.proto)
+
+## 4.1. Unary RPC
+
+The RPC service definition of a unary call is as follows.
+```proto
+service RouteGuide {
+    rpc GetFeature(Point) returns (Feature) {}
+}
+```
+The Ballerina service implementation of a gRPC can be done in two ways.
+1. Using direct returning
+2. Using a caller
+
+Directly returning the response is the most convenient implementation. However, for asynchronous RPC calls, directly returning is not suitable, and for such use cases, using a caller is the ideal approach. In addition, each RPC call (unary, server streaming, client streaming, and bidirectional streaming) can be implemented in both ways.
+
+**RPC Using Direct Return**
+
+Ballerina CLI generates the relevant service skeleton, and the implementation of the unary RPC call using direct return is as follows.
+
+```ballerina
+service "RouteGuide" on new grpc:Listener(8980) {
+
+    remote function GetFeature(Point point) returns Feature|error {
+        foreach Feature feature in FEATURES {
+            if feature.location == point {
+                return feature;
+            }
+        }
+        return {location: point, name: ""};
+    }
+}
+```
+
+Here, the RPC implementation creates a featured record and directly return it from the remote method.
+
+**RPC Using a Caller**
+
+The Ballerina implementation of the same unary RPC using a caller is as follows.
+
+```ballerina
+service "RouteGuide" on new grpc:Listener(8980) {
+
+    remote function GetFeature(RouteGuideFeatureCaller caller, Point point) returns error? {
+        Feature?|error feature = featureFromPoint(point);
+        if feature is Feature {
+            check caller->sendFeature(feature);
+        } else if feature is error {
+            check caller->sendError(<grpc:Error> feature);
+        } else {
+            check caller->sendFeature({location: {latitude: 0, longitude: 0}, name: ""});
+        }
+    }
+}
+```
+
+## 4.2. Server Streaming RPC
+The RPC service definition of a server streaming call is as follows.
+```proto
+service RouteGuide {
+    rpc ListFeatures(Rectangle) returns (stream Feature) {}
```

Comment:
```suggestion
    rpc ListFeatures(Rectangle) returns (stream Feature) {};
```

Reply:
Semicolon is not needed https://github.com/grpc/grpc-go/blob/v1.38.0/examples/route_guide/routeguide/route_guide.proto